[Peh Kai Min, Ryan] iP

ryanpeh opened the pull request:
# Duck 🦆

> “The perils of duck hunting are great - especially for the duck." – Walter Cronkite

Duck🦆 frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ _SUPER_ FAST to use ⚡️⚡️⚡️

All you need to do is,
1. download it from [here](https://github.com/ryanpeh/ip/releases).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you :wink:

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming eventually)

---

If you're a Java programmer(why?), you can use it to practice Java too (but why would you). Here's the `main` method:
```java{.no-line-numbers}
public class Duke {
    public static void main(String[] args) {
        new Duke("data/tasks.txt").run();
    }
}

Commit:
Add Gradle support

ryanpeh committed:
add greet, echo and exit functionality

ryanpeh committed:
add storing and displaying of tasks

ryanpeh committed:
add ability to mark task as done

ryanpeh committed:
Add support for tracking todos, events and deadlines

ryanpeh committed:
Add automated text UI testing

ryanpeh committed:
Refactor implementation of adding various tasks

ryanpeh committed:
Add override statements

ryanpeh committed:
Add error handling functionality

ryanpeh committed:
Add support for deleting tasks from list

ryanpeh committed:
Add storing of date and time as dateTime

ryanpeh committed:
Add saving of tasks in a file

ryanpeh committed:
Merge pull request #1 from ryanpeh/branch-Level-7

Add saving of tasks in a file

ryanpeh committed:
Merge pull request #2 from ryanpeh/branch-Level-8

Add storing of date and time as dateTime

ryanpeh committed:
Refactor code to be more OOP-like

ryanpeh committed:
Add JUnit tests

ryanpeh committed:
Add manifest

ryanpeh committed:
Refactor imports to follow coding standard

ryanpeh committed:
Add javadocs

ryanpeh committed:
Refactor imports to follow coding standard

ryanpeh committed:
Add ability to find task by searching for a keyword

ryanpeh committed:
Merge pull request #3 from ryanpeh/branch-A-JavaDoc

Add javadocs

ryanpeh committed:
Merge pull request #4 from ryanpeh/branch-A-CodingStandard

Refactor imports to follow coding standard

ryanpeh committed:
Merge branch 'master' into branch-Level-9

ryanpeh committed:
Merge pull request #5 from ryanpeh/branch-Level-9

Add ability to find task by searching for a keyword

ryanpeh committed:
Fix error during merging

ryanpeh committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

ryanpeh committed:
Change main class name

ryanpeh committed:
Add CheckStyle

ryanpeh committed:
Fix checkstyle issues

aryagiat commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,59 @@
+            throw new DukeException("Invalid Command Entered");
+        }
+        String command = matcher.group("command");
+        String arguments = matcher.group("arguments");
```
Should this variable be in plural form?

aryagiat commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+ */
+public class Storage {
+    private static final Pattern STORE_FORMAT = Pattern.compile("\\[(?<type>\\S)] \\[(?<done> |X)] (?<arguments>.*)");
+    private final String fileName;
```
should a constant variable be in cammelcase?

ryanpeh replied:
Constant names must be all uppercase using underscore to separate words, as mentioned in the java coding standard.

beefham commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+                }
+                taskList.addTask(task);
+                break;
+            default:
```
Minor nitpick: maybe you could use the default case to throw an exception again if needed.

ryanpeh replied:
Fixed, thanks!

yeppog commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,110 @@
+                break;
+            default:
+            }
+
```
perhaps the spaces here could be removed unless they serve some purpose?

ryanpeh replied:
Fixed, thanks!

yeppog commented on src/main/java/duke/commands/EventCommand.java:
```diff
@@ -0,0 +1,51 @@
+        if (!matcher.matches()) {
+            throw new DukeException("Events must be in the format 'event <description> /at <dd MMM yyyy HH:mm>");
+        }
+        description = matcher.group("description").trim();
```
might be nitpicking but would using the `this` keyword help to convey that the variable that is being assigned is an instance variable?

ryanpeh replied:
Actually in the advanced coding standard, it recommends only using the `this` keyword only when a field is shadowed by a method or constructor parameter.

yeppog commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+     */
+    public Deadline(String description, String by) {
+        super(description);
+        this.by = LocalDate.parse(by, formatter);
```
for some constructors, `this` is utilised but not in all, perhaps we could standardise this?

ryanpeh replied:
In this case, the `this` is necessary, therefore I didn't remove it.

yeppog commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+ */
+public class Deadline extends Task {
+    protected final LocalDate by;
+    private DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd MMM yyyy");
```
could possibly be declared `final` since we do not mutate its value.

ryanpeh replied:
Fixed, thanks!

ryanpeh committed:
Add GUI

ryanpeh committed:
Fix to work with GUI

ryanpeh committed:
Fix checkstyle and remove unused methods

yeppog reviewed (commented):
Good job!

ryanpeh committed:
Fix various coding standard issues

ryanpeh committed:
Add assertions

ryanpeh committed:
Improve code quality based on coding standard

Some areas of the code had improper and poor code comments, or unsafe coding
practices used.

Such low quality code can result in bugs, as well as increased difficulty
in maintaining and extending the codebase.

We should ensure that all code follows the coding standard so that it is
easier for others to work on the same codebase in the future.

ryanpeh committed:
Merge pull request #6 from ryanpeh/branch-A-Assertions

Add assertions

ryanpeh committed:
Merge branch 'master' into branch-A-CodeQuality

ryanpeh committed:
Add reminder for tasks occuring within the next 7 days

ryanpeh committed:
Add showing of reminder at startup

ryanpeh committed:
Merge pull request #7 from ryanpeh/branch-B-Reminders

Merge branch-B-Reminders into master

ryanpeh committed:
Change look of ui

ryanpeh committed:
Modify various ui display text

ryanpeh committed:
Merge pull request #9 from ryanpeh/branch-A-BetterGui

Branch a better gui

ryanpeh committed:
Add user guide

ryanpeh committed:
Add command to clear all tasks

ryanpeh committed:
Change user icons

ryanpeh committed:
Change displayed text when no tasks in list

ryanpeh committed:
Fix issue when adding deadlines and events

ryanpeh committed:
Add product screenshot

ryanpeh committed:
Fix linking of product screenshot in user guide

ryanpeh committed:
Fix checkstyle issues

ryanpeh committed:
Fix javafx loading issue

ryanpeh committed:
Update task file location

ryanpeh committed:
Fix clear command not clearing tasklist file

ryanpeh committed:
Fix checkstyle issues

ryanpeh committed:
Fix issue in creation of task file

ryanpeh committed:
Set theme jekyll-theme-tactile
